### ISSUE TITLE
"Split arrays into separate tables" is there are less than 5 elements in array

### ISSUE THREAD

**Issue by sorenabell** [open-contracting/spoonbill-web#306]

**Comment by sorenabell** [comment]
@jpmckinney @sabahfromlondon this rule seems more logical and is not difficult to implement.
Let us know whether we should change this rule for splitting tables.

Also, whether **5 and fewer elements** or just **fewer than 5 elements** should cause the split.

**Comment by jpmckinney** [comment]
> this rule seems more logical and is not difficult to implement

To clarify, can you express the rule?

> Also, whether 5 and fewer elements or just fewer than 5 elements should cause the split.

Which one is currently implemented?

**Comment by sabahfromlondon** [comment]
@jpmckinney for context what is happening in the flatten tool is that if there are fewer than 5 rows in a table then these are automatically rolled up into the main table - we specified that as a requirement. 

However, on the interface the user is told that there is an array, but there is no option to split out the array. I think to a user, this is confusing and might even look like a bug.

We have two options:
1. We keep the fields rolled up, but need to provide the user with the option to split out the fields via the toggle switch. 
2. We have the same behaviour for all arrays no matter now many fields in the array and automatically split out all tables.

My suggestion is to go with the first option.

**Comment by sorenabell** [comment]
@jpmckinney  The current rule is "Split arrays into separate tables if there are 5 or more elements in the array".
But you mentioned that you want the opposite to be happening.

**Comment by jpmckinney** [comment]
@sorenabell And from your comment, which rule seems more logical: the current rule (never split if < 5, offer to split if >= 5) or the other rule (never split if >=5, offer to split if < 5)?

**Comment by sabahfromlondon** [comment]
@jpmckinney 

> @jpmckinney The current rule is "Split arrays into separate tables if there are 5 or more elements in the array".
> But you mentioned that you want the opposite to be happening.

I'm just trying to address the issue with the UI which is confusing. Telling the user that there is an array in the table and then no other information, is likely to lead to confusion and potentially the user thinking something is wrong with the tool.

I guess a third option is just to add an extra sentence for this instance to say: "The array has fewer than five fields, so has not been split out into a separate table."

By the way @sorenabell the sentence "There is one array in this tables" should be "table" when there is only one array. 

<img width="875" alt="Screenshot 2021-07-21 093512" src="https://user-images.githubusercontent.com/11212467/126458460-32b032a8-be15-4d79-a686-dac6e6c732b5.png">

**Comment by jpmckinney** [comment]
Yes, we can clarify the language. I also just want to understand which behavior we actually want.

**Comment by sorenabell** [comment]
@sabahfromlondon This small text issue was fixed.

**Comment by jpmckinney** [comment]
From our call today, the new logic should be:

* Never merge a child table into a parent table if any parent object has >= [threshold] child objects
* Offer the option to merge a child table into a parent table if all parent objects have if < [threshold] child objects

Regarding the choice of operators (>= and < versus > and <=), we can continue using the same operators we currently do. From my reading, we do `should_split = len(item) >= self.table_threshold`

This should address one of the causes of #303, where the problem was that merging the child table would cause the creation of a very wide table.

For now, the default option can be that all child tables are split. (In future, we might want to test whether users prefer to have single-value children merged into parent tables by default.)

We also agreed that, on each screen (i.e. for each parent table), the split behavior will be all-or-nothing: that is, either all child tables are merged, or all child tables are split. Per-child-table controls are expected to be too complicated for the user. (In future, we might want to test whether users prefer to have per-child-table controls, like when removing tables.)

**Comment by jpmckinney** [comment]
To clarify, regarding the all-or-nothing implementation, if one child table A is above the threshold, and the other child tables (B, C) are below the threshold, what happens?

1. The user has no choice. Since one table is above the threshold, all child tables must be split out.
2. The user can decide whether B and C are both merged, or both split out. A is always split out.

**Comment by sorenabell** [comment]
The second option seems to be correct:
- The user can decide whether B and C are both merged, or both split out. A is always split out.

**Comment by jpmckinney** [comment]
Great!

**Comment by sorenabell** [comment]
@jpmckinney just re-confirming and re-phrasing once more: 

All child tables should be split (regardless of the number of elements in the array). 

- If there are fewer than [threshold] elements, then we should offer the merge/unsplit option.
- If there are more than [threshold] elements, then we should not offer the merge at all and leave all the child tables split.

Is that correct?

**Comment by jpmckinney** [comment]
Yes. The second bullet should be "equal to or greater than", to cover the "at threshold" case.

**Comment by sorenabell** [comment]
Wireframes: https://whimsical.com/ocp-ft-split-tables-Sf2pYFG2vYNjS8hJPUdVJw

**Comment by sorenabell** [comment]
@jpmckinney Should CLI mimic the splitting behavior discussed in this issue?

**Comment by jpmckinney** [comment]
If the CLI can be more flexible, it would be better.

Otherwise, the CLI should behave the same as the web version.

**Comment by sorenabell** [comment]
I've created a separate issue for CLI https://github.com/open-contracting/spoonbill/issues/176

**Comment by sorenabell** [comment]
The UI for the new splitting behavior was updated on the [staging](https://ocp12.open-contracting.org/) website. Please check.

**Comment by jpmckinney** [comment]
@sabahfromlondon Can you check this issue?
